crates/ruff/src/rules/flake8_tidy_imports/rules/banned_module_level_imports.rs

```diff
@@ -0,0 +1,108 @@
+use ruff_python_ast::{Ranged, Stmt};
+
+use ruff_diagnostics::{Diagnostic, Violation};
+use ruff_macros::{derive_message_formats, violation};
+use ruff_source_file::Locator;
+use ruff_text_size::TextRange;
+
+use crate::checkers::ast::Checker;
+
+/// ## What it does
+/// Checks for banned imports at module level. The banned imports are allowed inline, such as
+/// within a function definition or an `if TYPE_CHECKING:` block.
+///
+/// ## Why is this bad?
+/// Some modules take a relatively long time to import, such as `torch` or `tensorflow`. Library
+/// authors might want to ensure that you only pay the import cost for these modules if you
+/// directly use them, rather than if you import a module that happens to use an expensive module
+/// in one of its functions.
+///
+/// ## Options
+/// - `flake8-tidy-imports.banned-module-level-imports`
+#[violation]
+pub struct BannedModuleLevelImports {
+    name: String,
+}
+
+impl Violation for BannedModuleLevelImports {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let BannedModuleLevelImports { name } = self;
+        format!("`{name}` is banned at module level. Please move the import inline.")
+    }
+}
+
+enum NameMatchPolicy {
+    ExactOnly,
+    ExactOrParents,
+}
+
+fn banned_at_module_level_with_policy(
+    checker: &mut Checker,
+    name: &str,
+    stmt: &Stmt,
+    text_range: TextRange,
+    locator: &Locator,
+    policy: &NameMatchPolicy,
+) {
+    if !locator.is_at_start_of_line(stmt.start()) {
+        return;
+    }
```

Comment:
I think more robust here would be to use `checker.semantic().at_top_level()`, which will base its decision on scoping and parent statements. It's possible to have statements that aren't at the start of the line, but are at the top level, like:
```python
x = 1; import tensorflow
```

Reply:
Thanks! I shamelessly copied this from the `module_import_not_at_top_of_file` rule (`E402`), so I think that will need to be fixed too.

Reply:
Thanks for pointing that out, good call! Fixed in https://github.com/astral-sh/ruff/pull/6526/files.